samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java

```diff
@@ -0,0 +1,123 @@
+        super(BookingDialog.class.getName());
+
+        addDialog(new TextPrompt(TextPrompt.class.getName()));
+        addDialog(new ConfirmPrompt(ConfirmPrompt.class.getName()));
+        addDialog(new DateResolverDialog(null));
+        WaterfallStep[] waterfallSteps = {
+            this::destinationStep,
+            this::originStep,
+            this::travelDateStep,
+            this::confirmStep,
+            this::finalStep
+        };
+        addDialog(new WaterfallDialog(WaterfallDialog.class.getName(), Arrays.asList(waterfallSteps)));
+
+        // The initial child Dialog to run.
+        setInitialDialogId(WaterfallDialog.class.getName());
+    }
+
+    private CompletableFuture<DialogTurnResult> destinationStep(WaterfallStepContext stepContext) {
+        BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();
+
+        if (bookingDetails.getDestination() == null) {
+            Activity promptMessage =
+                MessageFactory.text(destinationStepMsgText, destinationStepMsgText, InputHints.EXPECTING_INPUT);
+            return stepContext.prompt(TextPrompt.class.getName(), new PromptOptions() {{
+                setPrompt(promptMessage);
+            }});
+        }
+
+        return stepContext.next(bookingDetails.getDestination());
+    }
+
+    private CompletableFuture<DialogTurnResult> originStep(WaterfallStepContext stepContext) {
+        BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();
+
+        bookingDetails.setDestination(stepContext.getResult().toString());
+
+        if (StringUtils.isNotBlank(bookingDetails.getOrigin())) {
+            Activity promptMessage =
+                MessageFactory.text(originStepMsgText, originStepMsgText, InputHints.EXPECTING_INPUT);
+            return stepContext.prompt(TextPrompt.class.getName(), new PromptOptions() {{
+                setPrompt(promptMessage);
+            }});
+        }
+
+        return stepContext.next(bookingDetails.getOrigin());
+    }
+
+    private CompletableFuture<DialogTurnResult> travelDateStep(WaterfallStepContext stepContext) {
+        BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();
+
+        bookingDetails.setTravelDate(stepContext.getResult().toString());
+
+        if (StringUtils.isNotBlank(bookingDetails.getTravelDate()) || isAmbiguous(bookingDetails.getTravelDate())) {
+            return stepContext.beginDialog(DateResolverDialog.class.getName(), bookingDetails.getTravelDate());
+        }
+
+        return stepContext.next(bookingDetails.getTravelDate());
+    }
+
+    private CompletableFuture<DialogTurnResult> confirmStep(WaterfallStepContext stepContext) {
+        BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();
+
+        bookingDetails.setTravelDate(stepContext.getResult().toString());
+
+        String messageText =
+            String.format("Please confirm, I have you traveling to: %1s from: %2s on: %3s. Is this correct?",
```

Comment:
```suggestion
            String.format("Please confirm, I have you traveling to: %s from: %s on: %s. Is this correct?",
```

If we use the parameters by order, it's not necessary to specify the index